File: Assets/Scripts/datastructs/States.cs
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using UnityEngine;
using Debug = UnityEngine.Debug;

public class State {
    public int id;
    public bool complete;

    public State(int? _id, bool _complete=false) {
        (id, complete) = (_id ?? -4, _complete);
    }

    public static State LOST = new(-1, true);
    public static State CONFUSED = new(-2, true);
    public static State ERROR = new(-3, true);
    public static State UNKNOWN = new(-4, true);
    public static State ORPHAN = new(-5, true);

    public bool IsValid() {
        return IsReal() || id == -5;
    }

    public bool IsReal() { return id >= 0; }

    public override string ToString() {
        return $"State({id},{complete})";
    }
}

public class Cycle {
    private List<int> values;
    private int cur;

    public Cycle(List<int> values) {
        this.values = values;
        cur = 0;
    }

    public void Incr() {
        cur = (cur + 1) % values.Count;
    }

    public int Get(bool incr = true) {
        int v = values[cur];
        if (incr) Incr();
        return v;
    }
}
public class Group : List<List<int>> {}

public class ConfluenceDict : Dictionary<string,Dictionary<int,Group>> {
    public override string ToString() {
        return string.Join("\n", this.Select(p => $"({p.Key} : {string.Join("\n", p.Value.Select(k => $"({k.Key} : [{string.Join(",", k.Value.Select(l => $"{string.Join(",", l)}"))}])"))})"));
    }
}

public class StateRules {
    public VirtualNetwork network;
    public Dictionary<string, Cycle> producerCycles;

    public StateRules(VirtualNetwork _network) {
        network = _network;
        producerCycles = GenerateProducerCycles();
    }

    public Dictionary<string,Cycle> GenerateProducerCycles() {
        var producers = network.GetProducers();
        Dictionary<string, List<int>> idDict = new();
        producers.ForEach(p => {
            idDict.TryAdd(p.name, new());
            idDict[p.name].Add(p.id);
        });
        return idDict.ToDictionary(p => p.Key, p => new Cycle(p.Value));
    }

    public State Assign(string producer) {
        if (producer == null) return State.ORPHAN;
        return new(producerCycles.GetValueOrDefault(producer)?.Get(), true);
    }

    public State Update(State state, string selected=null) {
        if(!state.complete) return state;
        if(!state.IsValid()) return State.ERROR;
        if(!state.IsReal()) {
            var node = network.nodes.Where(n => n.name.Equals(selected) && n.isDefault).FirstOrDefault();
            return node != default(VNNode) ? new(node.id) : State.ERROR;
        }
        if(!network.HasNode(state.id)) return State.ERROR;
        var neighbours = network.GetNeighbours(state.id);
        if(neighbours.Count == 1) return new(neighbours[0].id);
        if(selected==null) return State.CONFUSED;
        var choice = neighbours.Where(n => n.name.Equals(selected)).First();                 //neighbours
        choice ??= network.nodes.Where(n => n.isDefault && n.name.Equals(selected)).First(); //default nodes
        return choice == null ? State.LOST : new(choice.id);
    }

    public string GetCurrentService(State state) {
        if(!state.IsReal()) return null;
        return network.GetNode(state.id).name;
    }

    public List<string> GetPotentialServices(State state) {
        if(!state.IsReal()) return new();
        return network.GetNeighbours(state.id).Select(n => n.name).ToList();
    }

    public List<string> GetDefaultServices() {
        return network.nodes.Where(n => n.isDefault).Select(n => n.name).ToList();
    }

    public ConfluenceDict GenerateConfluenceDict() {
        ConfluenceDict cd = new();
        foreach(var n in network.GetConsumers()) {
            cd.TryAdd(n.name, new());
            cd[n.name][n.id] = new();
            foreach(var pn in network.GetPreNeighbours(n.id)) {
                var edge = network.edgeDict[(pn.id, n.id)];
                var list = cd[n.name][n.id].Find(l => edge.SameGroup(network.edgeDict[(l[0], n.id)].group));
                if(list == default(List<int>)) cd[n.name][n.id].Add(new(){ pn.id });
                else list.Add(pn.id);
            }
        }
        return cd;
    }
}

File: Assets/Python/root/datastructs/states.py
import networkx as nx
from networkx import DiGraph
import matplotlib.pyplot as plt
from collections import defaultdict

from util.json_helper import read_json_file

class State:
    def __init__(self, id: int, complete: bool=False):
        self.id = id
        self.complete = complete

    #Selected an invalid direction when path split
    def lost():
        return State(-1, True)
    #Didn't select a direction when path split
    def confused():
        return State(-2, True)
    #Selected a service that doesn't exist
    def error():
        return State(-3, True)
    #State doesn't exist
    def unknown():
        return State(-4, True)
    #state was made without unknown parents
    def orphan():
        return State(-5, True)
    
    def is_valid(self):
        return self.is_real() or self.id == -5
    
    def is_real(self):
        return self.id >= 0
    
    def __str__(self) -> str:
        return f"({self.id},{self.complete})"

class StateGraph(DiGraph):
    def from_virtual(virtual_network):
        G = StateGraph()
        for s in virtual_network["nodes"]:
            G.add_node(s['id'], label=s['name'], default=s['default'] if 'default' in s else False)
        for c in virtual_network["edges"]:
            G.add_edge(c['a'], c['b'], label=c['type'], group=c['group'] if 'group' in c else None)
        return G
    
    def to_json(G, file_name: str):
        raise NotImplementedError("Function not implemented")
    
    def save_plot(G, file_name: str):
        pos = nx.spring_layout(G)  # You can choose a different layout if needed
        labels = {node: data['label'] for node, data in G.nodes(data=True)}
        edge_labels = {(u, v): data['label'] for u,v, data in G.edges(data=True)}
        plt.figure()
        nx.draw(G, pos, with_labels=True, labels=labels, node_size=1000, node_color='skyblue', font_size=8, font_color='black', font_weight='bold', arrowsize=20)
        nx.draw_networkx_edge_labels(G, pos, edge_labels=edge_labels, font_color='red')

        plt.savefig(file_name)

    def get_sources(G):
        return [n for n in G.nodes() if G.in_degree(n) == 0]
    
    def get_consumers(G):
        return [n for n in G.nodes() if G.in_degree(n) > 0]

class Cycle:
    def __init__(self, values: list[int]) -> None:
        self.length = len(values)
        self.values = values
        self.cur = 0

    def incr(self):
        self.cur += 1
        if self.cur >= self.length:
            self.cur = 0

    def get(self, incr=True) -> int:
        v = self.values[self.cur]
        if incr: self.incr()
        return v

class StateRules:
    def __init__(self, graph: StateGraph) -> None:
        self.graph = graph
        self.producer_cycles = self.generate_prod_cycles()

    def generate_prod_cycles(self) -> dict[str, Cycle]:
        producers = [(n, self.graph.nodes[n]['label']) for n in self.graph.get_sources()]
        id_dict = defaultdict(list)
        for id, name in producers: id_dict[name].append(id)
        return {n: Cycle(ids) for n,ids in id_dict.items()}

    def assign(self, producer: str) -> State:
        if not producer: return State.orphan()
        return State(self.producer_cycles[producer].get(), True)

    def update(self, state: State, selected: str = None, default_allowed: bool = False) -> State:
        if not state.complete: return state
        G = self.graph
        if not state.is_valid(): return State.error()
        if not state.is_real():
            if not default_allowed: return State.error()
            for n in G:
                if G.nodes[n]['label'] == selected and G.nodes[n]['default']:
                    return State(n)
        neighbours = G.neighbors(state.id)
        if G.degree(state.id) == 1: 
            return State(next(neighbours, None))
        if selected is None:
            return State.confused()
        for n in neighbours:
            if G.nodes[n]['label'] == selected: 
                return State(n)
        if default_allowed:
            for n in G:
                if G.nodes[n]['label'] == selected and G.nodes[n]['default']:
                    return State(n)
        return State.lost()
    
    def get_current_service(self, state: State) -> str:
        if not state.is_real(): return None
        return self.graph.nodes[state.id]['label']

    def get_potential_services(self, state: State) -> list[str]:
        if not state.is_real(): return []
        G = self.graph
        return [G.nodes[n]['label'] for n in G.neighbors(state.id)]
    
    def get_default_services(self) -> list[str]:
        G = self.graph
        return [G.nodes[n]['label'] for n in G if G.nodes[n]['default']]
    
    def generate_confluence_dict(self) -> dict[str,dict[int,list[list[int]]]]:
        G = self.graph
        node_to_groups = defaultdict(list)
        
        for u, v, data in G.edges(data=True):
            for l in node_to_groups[v]:
                for n in l:
                    other_group = G.edges[n, v]['group']
                    if other_group and other_group == data.get('group'):
                        l.append(u)
                        break
                else: continue
                break
            else: node_to_groups[v].append([u])
        print([(G.nodes[n]['label'], n) for n in G.get_consumers()])
        cd = defaultdict(dict)
        for n in G.get_consumers(): cd[G.nodes[n]['label']][n] = node_to_groups[n]
        return cd






